src/main/kotlin/org/rust/lang/core/types/ty/Ty.kt

```diff
@@ -79,21 +69,17 @@ val Ty.hasTyTypeParameters
 val Ty.hasTyProjection
     get(): Boolean = BitUtil.isSet(flags, HAS_TY_PROJECTION_MASK)
 
-fun <K, V : TypeFoldable<V>> Map<K, V>.substituteInValues(map: Substitution): Map<K, V> =
-    mapValues { (_, value) -> value.substitute(map) }
+val Ty.hasReLifetimeParameters
+    get(): Boolean = BitUtil.isSet(flags, HAS_RE_LIFETIME_PARAMETER_MASK)
 
-fun <K, V : TypeFoldable<V>> Map<K, V>.foldValues(folder: TypeFolder): Map<K, V> =
-    mapValues { (_, value) -> value.foldWith(folder) }
-
-fun Substitution.get(psi: RsTypeParameter): Ty? {
-    return get(TyTypeParameter.named((psi)))
-}
+val Ty.needToSubstitute
+    get(): Boolean = hasTyTypeParameters || hasReLifetimeParameters
 
-fun mergeFlags(element: BoundElement<*>): TypeFlags =
-    element.subst.values.fold(0) { a, b -> a or b.flags } or element.assoc.values.fold(0) { a, b -> a or b.flags }
+fun mergeFlags(element: BoundElement<*>): KindFlags =
```

Comment:
Probably this will be better like this:
```
fun mergeFlags(element: BoundElement<*>): KindFlags =
    element.subst.kinds.mergeFlags() or element.assoc.values.mergeFlags()

fun mergeFlags(kinds: List<Kind>): KindFlags = kinds.mergeFlags()

private fun Collection<Kind>.mergeFlags() = this.fold(0) { a, b -> a or b.flags }
```

This may cause some compilaton troubles, but you get the idea